### DISCUSSION TITLE
Is is possible to use express-session for cookie-based authentication?

### DISCUSSION THREAD

**Discussion by pedromsfernandes** [enisdenjo/graphql-ws#257] [Q&A] [answered]
I have a NestJS server and an Apollo client with `graphql-ws`. I am using cookie-based authentication for my queries and mutations, and now I also need to authenticate the user in my subscriptions. Problem is, in every guide I've seen, only token-based auth is used. I found [this issue](https://github.com/apollographql/subscriptions-transport-ws/issues/466) which provides a solution, however that library is not maintained and as far as I know, graphql-ws `onConnect` method does not provide a `ws` argument. Can `connectionParams` be used for this? Thanks!

**Comment by enisdenjo** [comment, marked as answer]
Cookies are automatically added to the headers of WebSocket HTTP upgrade requests by the browser.

Authenticate the client the same way you would a regular HTTP request. A great place to do auth is in the [`onConnect`](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/ServerOptions.md#onconnect) callback, you may find the HTTP upgrade request in the `ctx.extra.request`.

**Reply by pedromsfernandes** [reply]
That's what I was missing! Thank you :+1:

**Reply by Nickersoft** [reply]
I recently encountered this thread after having the same question 😅 I noticed though, that the `extra` field of the `onConnect` callback context is fully generic: https://github.com/enisdenjo/graphql-ws/blob/master/src/server.ts#L521. If `request` is always guaranteed to be there, could the type definitions be updated to reflect it? I feel a bit uneasy defining the type in my own code, as I'm not sure if the `request` will always be attached to the `extra` field.

**Reply by enisdenjo** [reply]
When you use one of the [`use/`](https://github.com/enisdenjo/graphql-ws/tree/master/src/use) servers, the context extra is typed and the documentation is [here](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/use_ws.Extra.md).

You can also [see how `makeServer` is used for `ws`](https://github.com/enisdenjo/graphql-ws/blob/master/src/use/ws.ts).

**Reply by Nickersoft** [reply]
Ah, got it – yeah I'm using `@nestjs/graphql` for my GraphQL setup, and it looks like their code [just plucks the `onConnect` type straight out of this package](https://github.com/nestjs/graphql/blob/master/packages/graphql/lib/services/gql-subscription.service.ts#L23), leaving the generic untyped. 

Actually, checking now, as of this writing, the [latest issue filed](https://github.com/nestjs/graphql/issues/2376) over on that repo mentions the lack of typing around `graphql-ws` extras, so I'll see if it leads to any movement over there. Thanks for the quick response!

**Comment by CemYil03** [comment]
Hey, I think I have quite a similar setup. In the background I use express-session but have no idea how to access the session object. I know that my session store executes the reuest correctly, but the response can't be found in the onConnect's context.
This is just the case for @Subscription. Mutations and Queries work perfectly.

**Reply by enisdenjo** [reply]
There is no response in the context, only the upgrade request - which is what you want because the [session is located in the request](https://github.com/expressjs/session#reqsession) for `express-session`.

The upgrade request can be found in [the `Context` extras](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/use_ws.Extra.md#request). For example:
```ts
import { WebSocketServer } from 'ws';
import { useServer } from 'graphql-ws/lib/use/ws';

const server = new WebSocketServer({
  port: 4000,
  path: '/graphql',
});

useServer(
  {
    // ...
    onConnect: (ctx) => {
      ctx.extra.request; 👈
    },
    onSubscribe: (ctx) => {
      ctx.extra.request; 👈
    },
    onNext: (ctx) => {
      ctx.extra.request; 👈
    },
    onError: (ctx) => {
      ctx.extra.request; 👈
    },
    onComplete: (ctx) => {
      ctx.extra.request; 👈
    },
  },
  server,
);
```

**Reply by CemYil03** [reply]
First: Thank you very much for the quick answer.

Here is where and how I inspect the context:

Console logging `ctx.extra.request.session` instead, returns undefined. The session cookie however which only contains the session id, but not the resolved session can be found in 2 or 3 places when logging `ctx.extra.request`.

Any idea?

<img width="723" alt="Bildschirmfoto 2022-07-11 um 21 29 37" src="https://user-images.githubusercontent.com/42110001/178343613-38734d4b-e293-4c90-b887-00e2e151f715.png">

**Reply by enisdenjo** [reply]
Very likely an ordering issue - if the request doesn't have the session, then `express-session` didn't run yet. You must make sure it runs before `graphql-ws`.

**Reply by CemYil03** [reply]
I think I kind of got it to work by placing the session middleware inside of onConnect. But one question that stays open for me is:
The session middleware also sets the session-cookie once a response is sent (so it would need to be able to operate on the response object).

In case of graphql-transport-ws queries and mutations came per regular http request instead of per websocket.
But because with graphql-ws for example the sign in mutation comes in and I can't think of a way how to initialize the session.
Any idea about that?

**Reply by enisdenjo** [reply]
> In case of graphql-transport-ws queries and mutations came per regular http request instead of per websocket.

`graphql-ws` (subprotocol `graphql-transport-ws`) is an WebSocket exclusive protocol and library - when using it, everything is transported over WS (queries, mutations and subscriptions).

If you want to run queries and mutations through HTTP, you have to make that split and not use the `graphql-ws` client.

> The session middleware also sets the session-cookie once a response is sent

You cannot set cookies from within an established WebSocket connection. You should instead use regular HTTP requests for the login mutation and set the cookies in the response as you would.

**Reply by CemYil03** [reply]
Okay thank you, I think this was the thing I had not wrapped my head around yet.
Maybe this would be something to point put in the documentation, so that others don't loose as much time with it.

Summarized: If someone wants only use graphql-ws for his GraphQL-Server and no other kinds of requests, then it is not possible to use session- / cookie-based authentication.

**Reply by CemYil03** [reply]
To resolve this it would nice to have a configuration option in the createClient method, which allows to still send queries and mutations via http requests and only handle subscriptions via the websocket connection

**Reply by enisdenjo** [reply]
> Maybe this would be something to point put in the documentation, so that others don't loose as much time with it.
>
> Summarized: If someone wants only use graphql-ws for his GraphQL-Server and no other kinds of requests, then it is not possible to use session- / cookie-based authentication.

Sessions around cookies is a HTTP feature, not WebSockets. WebSockets use HTTP **only** to perform a transport protocol upgrade, after the upgrade is done - all communication happens within the WebSocket connection and not through HTTP.

> ...configuration option in the createClient method, which allows to still send queries and mutations via http requests and only handle subscriptions via the websocket connection

No, this is a WebSocket library and transport. GraphQL over HTTP is a completely different transport which implements its own specification. You have to make a split when to use HTTP vs. WS yourself.
